ext/opcache/zend_file_cache.c

```diff
@@ -1714,6 +1715,21 @@ static void zend_file_cache_unserialize_class(zval                    *zv,
 		ZEND_MAP_PTR_INIT(ce->mutable_data, NULL);
 		ZEND_MAP_PTR_INIT(ce->static_members_table, NULL);
 	}
+
+	// Memory addresses of object handlers are not stable. They can change due to ASLR or order of linking dynamic. To
+	// avoid pointing to invalid memory we relink default_object_handlers with the first internal class in the class
+	// hierarchy, or pick the appropriate default handler if there is no internal parent class.
+	const zend_object_handlers *default_object_handlers = ce->ce_flags & ZEND_ACC_ENUM ? &zend_enum_object_handlers : &std_object_handlers;
+	zend_class_entry *parent = ce->parent;
+	if (parent && (ce->ce_flags & ZEND_ACC_LINKED)) {
+		while (parent && parent->type != ZEND_INTERNAL_CLASS) {
+			parent = parent->parent;
+		}
+		if (parent) {
+			default_object_handlers = parent->default_object_handlers;
+		}
+	}
+	ce->default_object_handlers = default_object_handlers;
```

Comment:
Oh. I guess it would be enough to set `default_object_handlers` to `std_object_handlers`/`zend_enum_object_handlers` then?

Reply:
Oh shit! I missed, when these ``ce->default_object_handlers`` were added :(
That separation of handlers from class entries was done by design to allow creation of proxy objects.
I don't see a big reason of adding that ``ce->default_object_handlers`` in 9e6eab3c139
I suspect, this is done for extending observer capabilities or to allow some run-time tricks by third-party extension.
@bwoebi ?

I'm not sure about all consequences of this change.
Personally, I would revert the original patch.

Reply:
@dstogov I think the rationale was that since different object handlers for instances of the same class are rarely used (or rather never, there wasn't a known case) storing it on each object doesn't really make sense. Removing `handlers` would reduce the size of `zend_object` from 56 to 48 bytes. On the other hand, it could increase cache misses when invoking handlers due to looking up handlers in the class (but probably unlikely, given that the class is used most of the time).

Reply:
Removing handlers from zend_object will break ability of proxy objects and some extensions (e.g. ext/ffi).

Also as you said, calling handlers through class_entry would require additional memory load that may cause cache miss.

Reply:
> Removing handlers from zend_object will break ability of proxy objects and some extensions (e.g. ext/ffi).

I see. I wasn't aware it was used in this way there. Having different handlers per object could be avoided by storing flags on the `CData` object itself instead and then `switch`ing on those in the given handler. That is if we think removing the handlers from the class is worth it. It probably doesn't make sense to have it in both places.

@bwoebi Let us know if there are any additional benefits to having the handlers on the class itself. It might also be worth it to check if there is any performance degradation or improvement (could really be either, compacting classes could mean we're potentially making better use of a given cache line).

Reply:
Removing breaks even bundled PHP extensions. I expect this also will break few third party extensions.
This kind of engine changes should be discussed before committing.

Reply:
> I suspect, this is done for extending observer capabilities or to allow some run-time tricks by third-party extension.

Yes. Both, actually. Currently, replacing of handlers requires you to get hold of some object instance including the handlers. It's sort of possible to achieve it, by instantiating an object in MINIT stage, then freeing it. But that's a terrible hack and not always without side-effects.

Reply:
Replacing object handers is a hack of course, and this may work improperly.

Did you try to set object->handlers through overridden zend_class_entry.create_object?
This should work without BC breaks, and prevent some unsafe optimizations and JIT-ing.

Reply:
@dstogov This gets complicated very quickly, if the object handlers turn out to be dynamic (like FFI CData) (or another extension doing similar tricks and co-existing with ours). To ensure proper handling then, I'd essentially have to store the old and new object handlers in run-time allocated structures, stored in a hashmap addressed via the object pointer, which then needs to be freed in a free_obj handler.
This adds quite the overhead.

Reply:
I didn't get how this is related. Nowadays zend_class_entries are usually stored in opcache SHM. They are immutable. So changing zend_class_entry.create_object or introducing zend_class_entry->default_handlers doesn't make a lot of difference. Both may be changed/set only at MINIT. Both cases won't affect run-time assignments of object->handlers in FFI::CData.